tests/mocks/core/services/request/RequestMock.php

```diff
@@ -0,0 +1,211 @@
+use EventEspresso\core\services\request\Request;
+
+/**
+ * Class RequestMock
+ * Description
+ *
+ * @package EventEspresso\tests\mocks\core\services\request
+ * @author  Brent Christensen
+ * @since   $VID:$
+ */
+class RequestMock extends Request
+{
+
+    /**
+     * $_GET parameters
+     *
+     * @var array $get
+     */
+    public $get;
+
+    /**
+     * $_POST parameters
+     *
+     * @var array $post
+     */
+    public $post;
+
+    /**
+     * $_COOKIE parameters
+     *
+     * @var array $cookie
+     */
+    public $cookie;
+
+    /**
+     * $_SERVER parameters
+     *
+     * @var array $server
+     */
+    public $server;
+
+    /**
+     * $_REQUEST parameters
+     *
+     * @var array $request
+     */
+    public $request;
+
+    /**
+     * @var RequestTypeContextCheckerInterface
+     */
+    public $request_type;
+
+    /**
+     * IP address for request
+     *
+     * @var string $ip_address
+     */
+    public $ip_address;
+
+    /**
+     * @var string $user_agent
+     */
+    public $user_agent;
+
+    /**
+     * true if current user appears to be some kind of bot
+     *
+     * @var bool $is_bot
+     */
+    public $is_bot;
+
+
+    public function __construct(array $get, array $post, array $cookie, array $server, $ip_address = '0.0.0.0')
+    {
+        $this->get = $get;
+        $this->post = $post;
+        $this->cookie = $cookie;
+        $this->server = $server;
+        $this->request = array_merge($this->get, $this->post);
+        $this->ip_address = $this->visitorIp($ip_address);
+        parent::__construct($get, $post, $cookie, $server);
+    }
+
+
+    /**
+     * @param string $key
+     * @param string $value
+     * @param bool   $override_ee
+     * @return    void
+     */
+    public function setRequestParam($key, $value, $override_ee = false)
+    {
+        $this->request[ $key ] = $value;
+    }
+
+
+    /**
+     * returns the value for a request param if the given key exists
+     *
+     * @param string $key
+     * @param null   $default
+     * @return mixed
+     */
+    public function getRequestParam($key, $default = null)
+    {
+        return isset($this->request[ $key ]) ? $this->request[ $key ] : $default;
+    }
+
+
+    /**
+     * check if param exists
+     *
+     * @param string $key
+     * @return bool
+     */
+    public function requestParamIsSet($key)
+    {
+        return isset($this->request[ $key ]);
+    }
+
+
+    /**
+     * check if a request parameter exists whose key that matches the supplied wildcard pattern
+     * and return the value for the first match found
+     * wildcards can be either of the following:
+     *      ? to represent a single character of any type
+     *      * to represent one or more characters of any type
+     *
+     * @param string     $pattern
+     * @param null|mixed $default
+     * @return false|int
+     */
+    public function getMatch($pattern, $default = null)
```

Comment:
Any tests against this method aren't really testing actual production code.  It's just testing this mock method.

Reply:
You wouldn't use the `RequestMock` class to test this method. The point of simplifying this method is to remove as much complexity and/or side effects and/or influence of the mocked class from your tests on logic that would normally utilize the mocked class.

Consider the following:

If we wanted to be absolutely sure that our tested logic was behaving as expected, then it would make most sent to be absolutely sure of what was being returned from any external functions our tested logic was using. This would be the ideal way to go about this:

```php
    // in RequestMock

    /**
     * @var mixed $get_match_response
     */
    private $get_match_response;
    
    /**
     * @param mixed $get_match_response
     */
    public function setGetMatchResponse($get_match_response)
    {
        $this->get_match_response = $get_match_response;
    }

    /**
     * @param string     $pattern
     * @param null|mixed $default
     * @return mixed
     */
    public function getMatch($pattern, $default = null)
    {
        return $this->get_match_response;
    }
```

Now we have absolute total control over what our tested class receives when it calls `RequestMock::getMatch()`.

Want to see what happens when our tested class receives `null` after calling `RequestMock::getMatch()` ?

```php
// in actual unit test method
$request_mock->setGetMatchResponse(null);
$result = $tested_class->someMethodThatUsesGetMatch();
// now assert that result is as expected
```

I didn't go that far because this requires more work and I felt that the current tests were sufficient, but if I wanted to be totally methodical/surgical/precise then something like the above where you specifically control the responses of your mocked methods is far superior.